File: examples/statlib/reference/statlib.ts
/**
 * The reference implementation: a small descriptive-statistics library that
 * stands in for "the codebase you are about to rewrite". The recorded corpus
 * in ../record.js drives this boundary; the port in ../port must match it.
 */

function assertNonEmpty(values: number[]): void {
  if (values.length === 0) throw new RangeError('statlib: input must not be empty');
}

function sorted(values: number[]): number[] {
  return [...values].sort((a, b) => a - b);
}

export function mean(values: number[]): number {
  assertNonEmpty(values);
  let sum = 0;
  for (const v of values) sum += v;
  return sum / values.length;
}

export function median(values: number[]): number {
  assertNonEmpty(values);
  const s = sorted(values);
  const mid = Math.floor(s.length / 2);
  return s.length % 2 === 1 ? s[mid]! : (s[mid - 1]! + s[mid]!) / 2;
}

/** Linear interpolation between closest ranks (the R-7 / NumPy default). */
export function percentile(values: number[], p: number): number {
  assertNonEmpty(values);
  if (!(p >= 0 && p <= 100)) {
    throw new RangeError('statlib: percentile p must be within [0, 100]');
  }
  const s = sorted(values);
  const rank = (p / 100) * (s.length - 1);
  const lo = Math.floor(rank);
  const frac = rank - lo;
  if (frac === 0) return s[lo]!;
  return s[lo]! + frac * (s[lo + 1]! - s[lo]!);
}

/** Sample standard deviation (Bessel's correction). */
export function stddev(values: number[]): number {
  if (values.length < 2) throw new RangeError('statlib: stddev needs at least two values');
  const m = mean(values);
  let sq = 0;
  for (const v of values) sq += (v - m) * (v - m);
  return Math.sqrt(sq / (values.length - 1));
}

export interface Summary {
  count: number;
  min: number;
  max: number;
  mean: number;
  median: number;
  p95: number;
  stddev: number | null;
}

export function summarize(values: number[]): Summary {
  assertNonEmpty(values);
  const s = sorted(values);
  return {
    count: values.length,
    min: s[0]!,
    max: s[s.length - 1]!,
    mean: mean(values),
    median: median(values),
    p95: percentile(values, 95),
    stddev: values.length >= 2 ? stddev(values) : null,
  };
}

/** Z-score normalization. Async to exercise the async timing class. */
export async function normalize(values: number[]): Promise<number[]> {
  const m = mean(values);
  const sd = stddev(values);
  await Promise.resolve();
  return values.map((v) => (v - m) / sd);
}

File: adapters/rust/examples/statlib.rs
//! A correct Rust port of examples/statlib/reference/statlib.js, served as a
//! doppel adapter. With the recorded contract from the statlib example:
//!
//!   cargo build --example statlib
//!   doppel check \
//!     --contract examples/statlib/contracts/statlib.dopl.jsonl \
//!     --adapter "./adapters/rust/target/debug/examples/statlib"
//!
//! Every operation mirrors the reference implementation's order of floating-
//! point operations, so results are bit-identical and the check reports all
//! interactions identical.

use doppel_adapter::{serve, Adapter, AdapterError};
use serde_json::{json, Value};

fn nums(v: &Value) -> Result<Vec<f64>, AdapterError> {
    let arr = v
        .as_array()
        .ok_or_else(|| AdapterError::new("TypeError", "expected an array of numbers"))?;
    arr.iter()
        .map(|x| {
            x.as_f64()
                .ok_or_else(|| AdapterError::new("TypeError", "expected an array of numbers"))
        })
        .collect()
}

fn assert_non_empty(values: &[f64]) -> Result<(), AdapterError> {
    if values.is_empty() {
        return Err(AdapterError::new(
            "RangeError",
            "statlib: input must not be empty",
        ));
    }
    Ok(())
}

fn sorted(values: &[f64]) -> Vec<f64> {
    let mut s = values.to_vec();
    s.sort_by(|a, b| a.partial_cmp(b).unwrap());
    s
}

fn mean(values: &[f64]) -> Result<f64, AdapterError> {
    assert_non_empty(values)?;
    let mut sum = 0.0;
    for v in values {
        sum += v;
    }
    Ok(sum / values.len() as f64)
}

fn median(values: &[f64]) -> Result<f64, AdapterError> {
    assert_non_empty(values)?;
    let s = sorted(values);
    let mid = s.len() / 2;
    if s.len() % 2 == 1 {
        Ok(s[mid])
    } else {
        Ok((s[mid - 1] + s[mid]) / 2.0)
    }
}

fn percentile(values: &[f64], p: f64) -> Result<f64, AdapterError> {
    assert_non_empty(values)?;
    if !(0.0..=100.0).contains(&p) {
        return Err(AdapterError::new(
            "RangeError",
            "statlib: percentile p must be within [0, 100]",
        ));
    }
    let s = sorted(values);
    let rank = (p / 100.0) * (s.len() as f64 - 1.0);
    let lo = rank.floor() as usize;
    let frac = rank - lo as f64;
    if frac == 0.0 {
        return Ok(s[lo]);
    }
    Ok(s[lo] + frac * (s[lo + 1] - s[lo]))
}

fn stddev(values: &[f64]) -> Result<f64, AdapterError> {
    if values.len() < 2 {
        return Err(AdapterError::new(
            "RangeError",
            "statlib: stddev needs at least two values",
        ));
    }
    let m = mean(values)?;
    let mut sq = 0.0;
    for v in values {
        sq += (v - m) * (v - m);
    }
    Ok((sq / (values.len() as f64 - 1.0)).sqrt())
}

fn summarize(values: &[f64]) -> Result<Value, AdapterError> {
    assert_non_empty(values)?;
    let s = sorted(values);
    let sd = if values.len() >= 2 {
        json!(stddev(values)?)
    } else {
        Value::Null
    };
    Ok(json!({
        "count": values.len(),
        "min": s[0],
        "max": s[s.len() - 1],
        "mean": mean(values)?,
        "median": median(values)?,
        "p95": percentile(values, 95.0)?,
        "stddev": sd,
    }))
}

fn normalize(values: &[f64]) -> Result<Value, AdapterError> {
    let m = mean(values)?;
    let sd = stddev(values)?;
    let out: Vec<f64> = values.iter().map(|v| (v - m) / sd).collect();
    Ok(json!(out))
}

struct Statlib;

impl Adapter for Statlib {
    fn name(&self) -> &str {
        "statlib-rust-port"
    }

    fn invoke(&mut self, boundary: &str, args: Vec<Value>) -> Result<Value, AdapterError> {
        let export = boundary.split('#').nth(1).unwrap_or("");
        let first = args.first().cloned().unwrap_or(Value::Null);
        match export {
            "mean" => Ok(json!(mean(&nums(&first)?)?)),
            "median" => Ok(json!(median(&nums(&first)?)?)),
            "percentile" => {
                let p = args
                    .get(1)
                    .and_then(|v| v.as_f64())
                    .ok_or_else(|| AdapterError::new("TypeError", "expected a number for p"))?;
                Ok(json!(percentile(&nums(&first)?, p)?))
            }
            "stddev" => Ok(json!(stddev(&nums(&first)?)?)),
            "summarize" => summarize(&nums(&first)?),
            "normalize" => normalize(&nums(&first)?),
            other => Err(AdapterError::new(
                "ReferenceError",
                &format!("no implementation for boundary export {other}"),
            )),
        }
    }
}

fn main() -> std::io::Result<()> {
    serve(Statlib)
}
